File: Problems/1048_Longest_String_Chain.cpp
#include<iostream>
#include<vector>
#include<string>
#include<algorithm>

using namespace std;

class Solution {
public:
    int longestStrChain(vector<string>& words) {
        sort(words.begin(), words.end(), compareByLength);
        vector<int> lens;
        for(int k=0; k<words.size(); k++){
            int curr_len = 1;
            for(int i=0; i<(int)words[k].length(); i++){
                string temp = words[k].substr(0, i) + words[k].substr(i+1);
                for(int j=0; j<k; j++){
                    if (temp == words[j]){
                        curr_len = max(curr_len, lens[j]+1);
                    }
                }
            }
            lens.push_back(curr_len);
        }
        return *max_element(lens.begin(), lens.end());
    }

private:
    static bool compareByLength(string s1, string s2){
        return s1.length()<s2.length();
    }
};
int main(){
    return 0;
}

File: Problems/1048_Longest_String_Chain.py
from typing import List

# A clearer solution
class Solution:
    def longestStrChain(self, words: List[str]) -> int:
        words.sort(key = lambda word: len(word))
        hash_map = {}
        for word in words:
            curr_length = 1
            for i in range(len(word)):
                string = word[:i] + word[i+1:]
                if string in hash_map:
                    curr_length = max(curr_length, hash_map[string]+1)
                hash_map[word] = curr_length
        return max(hash_map.values())


# class Solution:
#     def longestStrChain(self, words: List[str]) -> int:
#         words.sort(key = lambda word: len(word))
#         hash_map = {}
#         max_len = 1
#         for word in words:
#             flag = True
#             for i in range(len(word)):
#                 string = word[:i] + word[i+1:]
#                 if string in hash_map:
#                     hash_map[word] = hash_map[string] + 1
#                     max_len = max(max_len, hash_map[word])
#                     flag = False
#             if flag:
#                 hash_map[word] = 1
#         return max_len